x/gov/keeper/proposal.go

```diff
@@ -95,6 +95,38 @@ func (keeper Keeper) SubmitProposal(ctx sdk.Context, messages []sdk.Msg, metadat
 	return proposal, nil
 }
 
+// CancelProposal will cancel proposal before the voting period ends
+func (keeper Keeper) CancelProposal(ctx sdk.Context, proposalID uint64, proposer string) error {
+	proposal, ok := keeper.GetProposal(ctx, proposalID)
+	if !ok {
+		return sdkerrors.Wrapf(types.ErrProposalNotFound, "proposal is not found with %d id", proposalID)
+	}
+
+	// Check creator of the proposal
+	if proposal.Proposer != proposer {
+		return sdkerrors.Wrapf(types.ErrInvalidProposer, "invalid proposer %s", proposer)
+	}
+
+	// Check if proposal is active or not
+	if (proposal.Status != v1.StatusDepositPeriod) && (proposal.Status != v1.StatusVotingPeriod) {
+		return sdkerrors.Wrapf(types.ErrInactiveProposal, "%d", proposalID)
+	}
+
+	// Check proposal voting period is ended.
+	if proposal.VotingEndTime != nil && proposal.VotingEndTime.Before(ctx.BlockTime()) {
+		return sdkerrors.Wrapf(types.ErrInactiveProposal, "voting period is already ended for this proposal %d", proposalID)
+	}
+
+	// update the status to StatusCanceled
+	proposal.Status = v1.StatusCanceled
+	// set to store
+	keeper.SetProposal(ctx, proposal)
+	// insert the proposal into cancel proposal queue
+	keeper.InsertCanceledProposalQueue(ctx, proposalID)
```

Comment:
Am I right that if the proposer sends a tx that gets includes in block N, then the EnbBlocker of block N will run this and delete the canceled proposal. Which means that there's no way to query for a canceled proposal.

If that's correct, then why not just:
- Remove `StatusCanceled`
- Remove the CanceledProposalQueue
- Directly burn/send deposits on CancelProposal msg execution, don't wait for EndBlocker

The only reason I see why we would want a CanceledProposalQueue is if we wanted to keep a canceled proposal in state (to be queryable) for X amount of blocks more.

Reply:
https://github.com/cosmos/cosmos-sdk/pull/13010#issuecomment-1286503461

Reply:
>  I thought like if we do all the process inside msg execution then proposer has to wait for result, because we need do take care deposits and votes inside the process

Proposer doesn't need to wait for result if they use `--broadcast-mode sync|async`. And final execution time is the same whether it happens in DeliverTx or EndBlock.

One additional consideration is who pays for the gas of burning/returning deposits. Currently, in your design, the proposer pays little fees, just enough to write to the queue. In my proposed design, proposer would pay for the whole execution fee.

Reply:
IMO it boils down to a UX problem: if the proposer cancels a proposal, do we prefer:
1. That the proposal still be queryable for a certain period of time?
2. Or it's fine that it completely disappears from state?

IMO 2 is fine.

Reply:
It will there until EndBlock execute the CanceledProposalQueue

Reply:
I also don't think we need the queue. Let's:
* delete immediately
* create an event that proposal was cancelled.